### PR TITLE
refactor how map and call primitives are tagged

### PR DESCRIPTION
**Pull request by mattjj**
This seems a little less arcane. Includes the commit from #2828, but I figured I'd factor out this cleanup.

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]
I think it might make for slightly less overall code if you have `MapPrimitive` be a subtype of `CallPrimitive`, but LGTM either way.

File: jax/interpreters/ad.py
Comment (jekbradbury) [inline jax/interpreters/ad.py:224]: Thoughts on whether rules for HOPs (like the transpose rule, but also process_call/process_map) should treat `call_jaxpr` as just another part of `params` now?

Comment (mattjj) [reply jax/interpreters/ad.py:224]: Oh hmm. An alternative would be to create a special Eqn for these which has a jaxpr slot, right? That seems better...

Comment (mattjj) [reply jax/interpreters/ad.py:224]: Or, I guess, a special non-params slot on the primitive. Maybe that's what you were thinking too?

Comment (mattjj) [reply jax/interpreters/ad.py:224]: Ah, that doesn't make sense: there's only one instance of the primitive. Params are really things we bundle with primitives in a jaxpr eqn. So I think we'd need special eqns.

I'll give that a shot!

Comment (jekbradbury) [reply jax/interpreters/ad.py:224]: I just meant that George's vision was to make `call_jaxpr` just another param; that would work for transpose and translation (plus any other initial style transforms) but (of course!) not for final style transforms. So I think my comment was mostly off base.

**Comment by jekbradbury** [comment]
I'm not sure I like `staged_out` as a tag name here. I'm guessing it means "outer `JaxprTrace`s will stage out instances of this primitive in their entirety, rather than partially-evaluating their bodies" (right?)—but it took me a while to realize that was what this probably referred to. Maybe `always_staged_out` or `fully_stage_out`?

**Comment by mattjj** [comment]
I think it’s a property of the primitive, whether its impl executes in Python or if it’s staged out of Python. Nothing on its own to do with tracing. But those are the primitives for which we would hoist as much as possible into them, hence the interaction with StagingJaxprTrace.

File: jax/interpreters/ad.py
Comment (mattjj) [reply jax/interpreters/ad.py:224]: Yeah but I think it’s not ideal as a param because then we have some invariants to enforce/check everywhere rather than having those invariants built into the data structure.

**Comment by jekbradbury** [comment]
OK, that makes sense! Thanks for explaining.

File: jax/interpreters/ad.py
Comment (gnecula) [reply jax/interpreters/ad.py:224]: The reason I thought that putting call_jaxpr in params is that it makes it easier to explain the Jaxpr language to users, since the initial style primitives already have Jaxprs in params. Perhaps we can take all the subjaxprs from params and keep them separate, with only `cond` having two of them?

File: jax/core.py
Comment (gnecula) [inline jax/core.py:226]: This is a nice change

File: jax/interpreters/ad.py
Comment (mattjj) [reply jax/interpreters/ad.py:224]: That's a great point.

I talked with @gnecula a bit more this morning and he made the point that we could (and likely should) treat these higher-order primitives more uniformly. In particular, we can treat staged-out xla_call and xla_pmap like the other higher-order primitives and not need a special case in jaxpr interpreters for them. We had some more specific ideas but basically our conclusion was to try out some things and see how they looked.

**Comment by mattjj** [comment]
I had some follow-up changes in mind to try out, like either

1. factoring out `xla_call_jaxpr`/`pmap_call_jaxpr` primitives that work like the other initial-style higher-order primitives so that there's no need for a special case analysis in jaxpr interpreters, or else
2. cleaning up the representation / case analysis so that instead of relying on `extract_call_jaxpr` we have special `JaxprCallEqn` / `JaxprMapEqn` classes which have `call_jaxpr` and `call_jaxpr`/`mapped_invars` fields respectively.

I implemented 2 already (in an un-pushed commit), but want to try 1 as it now seems more promising.

Do you think we should merge this PR now-ish, and make more changes in follow-ups? Or fuse all such changes together? The latter might be slightly nicer for any jaxpr consumers, since they won't have to make multiple changes, but it also means making more changes at once (and increasing the chance I run out of fuel on this).

In any case, @gnecula if you have time I plan to wait for your review / LGTM, since I could really benefit from your advice here. Don't hold back: I'm happy to make changes!

**Comment by mattjj** [comment]
I'm going to replace this with a fresh PR!

**Pull request closed without merging**
